### DISCUSSION TITLE
Variables or command expansion automatically enclosed with single quotes

### DISCUSSION THREAD

**Discussion by krpors** [fish-shell/fish-shell#9376] [Q&A] [answered]
Hi,

Fish version 3.3.1

In short, I want this to work:

```fish
$ set fish_trace on
$ whiptail --notags --menu "Select a file"  20 80 10 (find -maxdepth 1 -type f -printf '"%f" "%f" ')
-> find -maxdepth 1 -type f -printf '"%f" "%f" '
> whiptail --notags --menu 'Select a file' 20 80 10 '"bla.pem" "bla.pem" '
Box options: 
	--msgbox <text> <height> <width>
	--yesno  <text> <height> <width>
	--infobox <text> <height> <width>
        ... output removed for brevity ...
```

As you can see, the `(find)` command is eventually expanded to `'"bla.pem" "bla.pem" '`, which is incorrect input for the `whiptail` command (it is enclosed in single quotes).

How can I 'unquote' the output of the command? 

I've also tried setting the output of the `find` command into a variable, and then use that, but with the same results.

**Comment by faho** [comment, marked as answer]
**Fish only splits command substitutions on newlines**

That's the important information here. That's your problem.

>As you can see, the (find) command is eventually expanded to '"bla.pem" "bla.pem" ', which is incorrect input for the whiptail command (it is enclosed in single quotes).

It isn't. It's incorrect input for whiptail, but it's not "enclosed in single quotes".

What `fish_trace` here is showing is the equivalent of the command that would be executed. It's a fictional commandline.

Your argument here is *as if* you had used `'"bla.pem" "bla.pem" '` on the commandline. What is actually passed, what whiptail receives, is therefore one string:

```
"bla.pem" "bla.pem" 
```

(with a trailing space)

What you need to figure out is what `whiptail` wants here. If you had the output hardcoded, how would you write it?

What whiptail expects is pairs of "tag" and "item". I.e. you would run this like

```fish
whiptail --notags --menu 'Select a file' 20 80 10 "tag1" "item1" "tag2" "item2"
```

And you're already trying this, both tag and item are supposed to be the filename. Only, you need to arrange for them to be separate arguments. As I've explained above, fish splits command substitution output into separate arguments on newlines - each *line* becomes a separate argument. It doesn't tokenize it, quotes have no meaning in the output (this is the same in bash - except that splits on spaces and tabs as well, which has issues for filenames with spaces).

So you need to adjust `find`s format string to print the item and the tag on separate lines, which you can do with `\n`: `-printf "%f\n%f\n"`.

Here:

```fish
whiptail --notags --menu "Select a file"  20 80 10 (find -maxdepth 1 -type f -printf '%f\n%f\n')
```

So `find` prints:

```
bla.pem
bla.pem
secondfile
secondfile
thirdfile
thirdfile
```

and this is turned into something like

```fish
whiptail --notags --menu "Select a file"  20 80 10 bla.pem bla.pem secondfile secondfile thirdfile thirdfile
```

which will make it happy.

**Comment by krpors** [comment]
You, my friend are a hero for giving such a detailed answer. Omitting the newline in the output was indeed the problem. I will be sure to read up a bit more on the substitution splitting. 

Thanks so much!
